Add Scheduling Configuration reference doc (#21437)
kubernetes/website: content/en/docs/reference/scheduling/config.md

---
title: Scheduler Configuration
content_type: concept
weight: 20
---

{{< feature-state for_k8s_version="v1.19" state="beta" >}}

You can customize the behavior of the `kube-scheduler` by writing a configuration
file and passing its path as a command line argument.

<!-- overview -->

<!-- body -->

## Minimal Configuration

A minimal configuration looks as follows:

```yaml
apiVersion: kubescheduler.config.k8s.io/v1beta1
kind: KubeSchedulerConfiguration
clientConnection:
  kubeconfig: /etc/srv/kubernetes/kube-scheduler/kubeconfig
```

## Profiles

A scheduling Profile allows you to configure the different stages of scheduling
in the {{< glossary_tooltip text="kube-scheduler" term_id="kube-scheduler" >}}.

> Each stage is exposed in a [extension point](#extension-points).

Reviewer: ```suggestion
Each stage is exposed in an [extension point](#extension-points).
```

[Plugins](#scheduling-plugins) provide scheduling behaviors by implementing one
or more of these extension points.

You can configure a single instance of `kube-scheduler` to run

> [multiple profiles](#multiple-profiles)

Reviewer: ```suggestion
[multiple profiles](#multiple-profiles).
```

---
title: Scheduler Configuration
content_type: concept
weight: 20
---

{{< feature-state for_k8s_version="v1.19" state="beta" >}}

> The `KubeSchedulerConfiguration` is a configuration API for `kube-scheduler`

Reviewer 2: ```suggestion
You can customize the behavior of the `kube-scheduler` by writing a configuration
file and passing its path as a command line argument.
```

(the name KubeSchedulerConfiguration isn't a Kubernetes resource, and mentioning it might confuse some readers.

Author: Done

that can be provided as a file via `--config` command line flag.

<!-- overview -->

<!-- body -->

## Minimal Configuration

A minimal configuration looks as follows:

```yaml
apiVersion: kubescheduler.config.k8s.io/v1beta1
kind: KubeSchedulerConfiguration
clientConnection:
  kubeconfig: /etc/srv/kubernetes/kube-scheduler/kubeconfig
```

> ## Upgrading from `v1alpha2` to `v1beta1` {#beta-changes}

Reviewer 2: I would move the “Upgrading from” to the end of this page, just before ` ## {{% heading "whatsnext" %}}`

Author: Done

---

> title: Scheduling Configuration

Reviewer 2: If renaming a page, remember to add a redirect.

Author: will do

Author: Can you refer me to an example of how to do a redirect?

Author: Done

content_template: templates/concept
weight: 20
---

## What is new

- `BindTimeoutSeconds` field in scheduler `KubeSchedulerConfiguration` is removed in v1beta1 version, please configure it via plugin args for `VolumeBinding`.
- `v1beta1` Scheduler `Extender` encoding is now strict, its `HTTPTimeout` is `metav1.Duration` and `enableHttps` is renamed to `enableHTTPS`.

> - `RequestedToCapacityRatioArgs` decoding is now strict

Reviewer 2: I recommend against mentioning “new” in headings, at least outside of release notes. Wording like this doesn't age well.

Author: Any guidance on how to format release notes instead?

Reviewer 2: Release notes? That's more a SIG Release specialty I think. I don't myself know how those are implemented.

Reviewer 2: By release notes, I mean that this information would have a better home on https://relnotes.k8s.io/ and in https://kubernetes.io/docs/setup/release/notes/

Author: We have a series of PRs that in theory require a release note each. It was suggested here https://github.com/kubernetes/kubernetes/pull/91580#issuecomment-637774454 that we accumulate the changes in a website page and point to them. This with the intention of having all the migration actions in a single place.

What do you think about naming the section: `Migrating from v1alpha2 to v1beta1`?

Reviewer 3: Yes, this would be better phrased "Changes between v1alpha2 and v1beta1"

Reviewer 3: Remember to reference field names as they appear in the config file (with the json tag names, not the go struct names)

Reviewer 2: OK, so:
- the release notes in [kubernetes/kubernetes](https://github.com/kubernetes/kubernetes) summarize the changes to scheduling for v1.19
- this PR adds more detailed information that those release notes can link to

Have I got that right?

If you want to add extra “What's new in scheduling” details that feel too big to go in a release note, would a blog post work?

Reviewer 3: a reference doc for migrating from v1alpha2 to v1beta1 doesn't seem like it belongs in a blog post... it's an enduring reference that we want in a fixed location as long as this config version is around

Reviewer 4: Hi @sftim I'm on the release notes team for 1.19. I agree with @liggitt . While informative, release notes should be a rather short, to the point item in the list of the user facing changes. The whole reference doc cannot be included in the release notes website but for completeness, I'd suggest adding a link to it in the release note (see [here](https://kubernetes.io/docs/setup/release/notes/#kube-apiserver) for an example).

Author: > Remember to reference field names as they appear in the config file 

Will do. I want to have agreement on the section first.
`Changes between v1alpha2 and v1beta1` sgtm.

cc @saschagrunert @puerco from sig-release

Reviewer 2: > Changes between v1alpha2 and v1beta1

That heading works for me.

   sort pending Pods in the scheduling queue. Exactly one queue sort plugin
   may be enabled at a time.
1. `PreFilter`: These plugins are used to pre-process or check information
   about a Pod or the cluster before filtering. They can mark a pod as
   unschedulable.

> 1. `Filter`: These plugins are the equivalent of Predicates in a scheduling

Reviewer: I will submit another PR finishing the `PostFilter` part.

   weighted scores sum.
1. `Reserve`: This is an informational extension point that notifies plugins
   when resources have been reserved for a given Pod. Plugins also implement an
   `Unreserve` call that gets called in the case of failure during or after
   `Reserve`.
1. `Permit`: These plugins can prevent or delay the binding of a Pod.
1. `PreBind`: These plugins perform any work required before a Pod is bound.
1. `Bind`: The plugins bind a Pod to a Node. Bind plugins are called in order
   and once one has done the binding, the remaining plugins are skipped. At
   least one bind plugin is required.
1. `PostBind`: This is an informational extension point that is called after
   a Pod has been bound.

For each extension point, you could disable specific [default plugins](#scheduling-plugins)
or enable your own. For example:

```yaml
apiVersion: kubescheduler.config.k8s.io/v1beta1
kind: KubeSchedulerConfiguration
profiles:
  - plugins:
      score:
        disabled:
        - name: NodeResourcesLeastAllocated
        enabled:
        - name: MyCustomPluginA
          weight: 2
        - name: MyCustomPluginB
          weight: 1
```

> You can use `*`  as name in the disabled array to disable all default plugins

Reviewer: ```suggestion
You can use `*` as name in the disabled array to disable all default plugins
```

   filtering phase. The scheduler will then select the node with the highest
   weighted scores sum.
1. `Reserve`: This is an informational extension point that notifies plugins

>    when resources have being reserved for a given Pod. Plugins also implement an

Reviewer 5: ```suggestion
   when resources have been reserved for a given Pod. Plugins also implement an
```

>    `Unreserve` call, that gets called in the case of failure during of after

Reviewer 5: ```suggestion
   `Unreserve` call, that gets called in the case of failure during or after
```

>    `Unreserve` call, that gets called in the case of failure during of after

Reviewer 6: ```suggestion
   `Unreserve` call that gets called in the case of failure during or after
```

   `Reserve`.
1. `Permit`: These plugins can prevent or delay the binding of a Pod.
1. `PreBind`: These plugins perform any work required before a Pod is bound.
1. `Bind`: The plugins bind a Pod to a Node. Bind plugins are called in order
   and once one has done the binding, the remaining plugins are skipped. At
   least one bind plugin is required.
1. `PostBind`: This is an informational extension point that is called after
   a Pod has been bound.

> 

Reviewer 5: for PreFilter above, can you update the text to indicate that PreFilter can return an unschedulable status indicating that all nodes are filtered for the pod being scheduled.

Author: I think the description already is enough for users of component config. Such detail could go in https://kubernetes.io/docs/concepts/scheduling-eviction/scheduling-framework/

Reviewer 5: I think it is not clear that it can act as an all-nodes Filter, but this is minor.

Author: See new commit

  Service across nodes.
  Extension points: `PreFilter`, `Filter`, `Score`.
  
### Multiple profiles

You can configure `kube-scheduler` to run more than one profile.
Each profile has an associated scheduler name and can have a different set of
plugins configured in its [extension points](#extension-points). For example:

```yaml
apiVersion: kubescheduler.config.k8s.io/v1beta1
kind: KubeSchedulerConfiguration
profiles:
  - schedulerName: default-scheduler
  - schedulerName: no-scoring-scheduler

>     plugins:

Reviewer: Let's disable `preScore` as well.

      score:
        disabled:
        - name: '*'

> ```

Reviewer 7: Would it make sense to explain this example? For example by adding a paragraph:

> The example above will configure `kube-scheduler` to run two schedulers: the default one and `no-scoring-scheduler` with the `score` extension point disabled for all plugins.

Author: Good suggestion

Reviewer 8: this is interesting, so are we saying we will be able to use the same scheduler binary to act as two different schedulers based on the scheduler field in the Pod ?

Reviewer 5: Correct.

  Service across nodes.
  Extension points: `PreFilter`, `Filter`, `Score`.
  
### Multiple profiles

> `kube-scheduler` can be configured to

Reviewer 2: nit (OK to merge without addressing this):
```suggestion
You can configure the `kube-scheduler` to
```

Author: Done
